conans/model/build_info.py

```diff
@@ -697,29 +700,32 @@ def check_component_requires(self, conanfile):
         direct_dependencies = [r.ref.name for r, d in conanfile.dependencies.items() if r.direct
                                and not r.build and not r.is_test and r.visible and not r.override]
 
-        for e in external:
+        for e, req_comp in external:
             if e not in direct_dependencies:
+                extra_msg = f" (required from component '{req_comp}')" if req_comp else ""
                 raise ConanException(
-                    f"{conanfile}: required component package '{e}::' not in dependencies")
+                    f"{conanfile}: required component package '{e}::'{extra_msg} not in dependencies")
         # TODO: discuss if there are cases that something is required but not transitive
+        external = [e[0] for e in external]
         for e in direct_dependencies:
             if e not in external:
                 raise ConanException(
                     f"{conanfile}: Required package '{e}' not in component 'requires'")
 
     @property
     def required_components(self):
-        """Returns a list of tuples with (require, component_name) required by the package
-        If the require is internal (to another component), the require will be None"""
+        """Returns a list of tuples with (require, component_name, dependee_component_name) required by the package
+        - If the require is internal (to another component), the require will be None
+        - If the requirement comes from the main cpp_info, the dependee_component_name will be None"""
         # FIXME: Cache the value
         # First aggregate without repetition, respecting the order
-        ret = [r for r in self._package.requires]
-        for comp in self.components.values():
+        ret = [(r, None) for r in self._package.requires]
+        for comp_name, comp in self.components.items():
             for r in comp.requires:
                 if r not in ret:
-                    ret.append(r)
+                    ret.append((r, comp_name))
         # Then split the names
-        ret = [r.split("::") if "::" in r else (None, r) for r in ret]
+        ret = [(*r[0].split("::"), r[1]) if "::" in r[0] else (None, *r) for r in ret]
```

Comment:
This is the main change - `required_components` now returns a 3-tuple. This is a draft because as far as I can see, this might be breaking behaviour as this function is public, so not too sure on the best approach.

Also note that as far as I can tell, the unpacking operator `*` was implemented in Python 3.5 so we're good to use it

Reply:
Public, but not documented apparently.
Still, it reads a bit risky, yes, not fully sure if worth the risk. Maybe a different approach like re-computing things when the check fails for better error messages (making it exclusively local to the error handling)?